src/main/kotlin/pl/org/pablo/slack/money/slack/SlackServiceImpl.kt

```diff
@@ -11,23 +14,38 @@ class SlackServiceImpl(
         private val moneyService: MoneyService,
         private val userService: UserService,
         private val chatService: ChatService,
-        private val slackUserService: SlackUserService
+        private val slackUserService: SlackUserService,
+        private val argumentParser: ArgumentParser
 ) : SlackService {
 
     private val dateFormat = DateTimeFormatter.ofPattern("dd-MM-yy hh:mm")
+    private val addArgumentParser = AddArgumentParser()
 
     override fun add(slackRequest: SlackRequest): String {
-        val text = slackRequest.text.replace(Regex("<@([aA-zZ0-9]+)(\\|[aA-zZ0-9]+)?>"), { "@${it.destructured.component1()}" })
-        val regex = Regex("@([aA-zZ0-9]+) ([0-9]+) ?(.*)?")
-        if (!text.matches(regex)) {
-            throw IllegalArgumentException("Wrong command $text")
-        }
-        val (_, user, value, description) = regex.find(text)!!.groupValues
-        moneyService.addMoney(AddDto(slackRequest.user_id, user, value.toInt(), description))
-        chatService.sendMessage(user, "Added new payment from ${slackRequest.user_name} with value: $value")
+        val arg = argumentParser.parse(addArgumentParser, slackRequest.text)
```

Comment:
There's a design patter for this kinds of things: https://sourcemaking.com/design_patterns/interpreter/java/1

Reply:
It would solve the repetition with canParse and parse.

Reply:
Hmm I'm not sure. `canParse` has one crucial use case that is problematic is when parsing below command - `@U1 10 @U2 20 desc`. Its can be splited into: `P1` `P2` `desc`. The issue is that we know when P1 or P2 ends, but we have no idea what is starting. If we finished `P1` then based on next string value we need to decide whether it's `P2` or `desc`. It's kinda we need to figure out what the `Operator` actually is.